UM/Settings/MachineManager.py

```diff
@@ -67,6 +71,46 @@ def getAllMachineVariants(self, machine_id):
 
         return variants
 
+    def getAllMachineMaterials(self, instance_id):
+        generic_materials = []
+        machine_materials = []
+
+        machine = self.findMachineInstance(instance_id)
+        if not machine:
+            return machine_materials
+            
+        machine_type = machine.getMachineDefinition().getProfilesMachineId()
+        machine_variant = machine.getMachineDefinition().getVariantName()
+
+        for profile in self._profiles:
+            profile_type = profile.getType()
+            #Filter out "partial" profiles
+            if profile_type == "material":
+                generic_materials.append(profile.getName())
+                continue
+
+            material = profile.getMaterialName()
+            if not material or material in machine_materials:
+                continue
+
+            profile_machine_type = profile.getMachineTypeId()
+            profile_machine_variant = profile.getMachineVariantName()
+            profile_machine_instance = profile.getMachineInstanceName()
+
+            if (profile_machine_instance and profile_machine_instance == instance_id) or \
+                    (profile_machine_variant and profile_machine_variant == machine_variant and profile_machine_type == machine_type) or \
+                    (profile_machine_type == machine_type):
```

Comment:
This warrants some documentation... Fairly ugly construct itself, too, but I wouldn't know a better one.